pam_kwallet.c

```diff
@@ -0,0 +1,823 @@
+        return PAM_IGNORE;
+    }
+
+    const char *password;
+    result = pam_get_item(pamh, PAM_AUTHTOK, (const void**)&password);
+
+    if (result != PAM_SUCCESS) {
+        pam_syslog(pamh, LOG_ERR, "%s: Couldn't get password %s", logPrefix,
+                   pam_strerror(pamh, result));
+        return PAM_IGNORE;
+    }
+
+    if (!password) {
+        pam_syslog(pamh, LOG_NOTICE, "%s: Couldn't get password (it is empty)", logPrefix);
+        //Asking for the password ourselves
+        result = prompt_for_password(pamh);
+        if (result != PAM_SUCCESS) {
+            pam_syslog(pamh, LOG_ERR, "%s: Prompt for password failed %s",
+                       logPrefix, pam_strerror(pamh, result)
+            );
+            return PAM_IGNORE;
+        }
+    }
+
+    //even though we just set it, better check to be 100% sure
+    result = pam_get_item(pamh, PAM_AUTHTOK, (const void**)&password);
+    if (result != PAM_SUCCESS || !password) {
+        pam_syslog(pamh, LOG_ERR, "%s: Password is not there even though we set it %s", logPrefix,
+                   pam_strerror(pamh, result));
+        return PAM_IGNORE;
+    }
+
+    if (password[0] == '\0') {
+        pam_syslog(pamh, LOG_NOTICE, "%s: Empty or missing password, doing nothing", logPrefix);
+        return PAM_IGNORE;
+    }
+
+    char *key = strdup(password);
+    result = pam_set_data(pamh, kwalletPamDataKey, key, cleanup_free);
+
+    if (result != PAM_SUCCESS) {
+        free(key);
+        pam_syslog(pamh, LOG_ERR, "%s: Impossible to store the password: %s", logPrefix
+            , pam_strerror(pamh, result));
+        return PAM_IGNORE;
+    }
+
+    //TODO unlock kwallet that is already executed
+    return PAM_IGNORE;
+}
+
+static int drop_privileges(struct passwd *userInfo)
+{
+    /* When dropping privileges from root, the `setgroups` call will
+    * remove any extraneous groups. If we don't call this, then
+    * even though our uid has dropped, we may still have groups
+    * that enable us to do super-user things. This will fail if we
+    * aren't root, so don't bother checking the return value, this
+    * is just done as an optimistic privilege dropping function.
+    */
+    setgroups(0, NULL);
+
+    //Change to the user in case we are not it yet
+    if (setgid (userInfo->pw_gid) < 0 || setuid (userInfo->pw_uid) < 0 ||
+        setegid (userInfo->pw_gid) < 0 || seteuid (userInfo->pw_uid) < 0) {
+        return -1;
+    }
+
+    return 0;
+}
+
+static void execute_kwallet(pam_handle_t *pamh, struct passwd *userInfo, int toWalletPipe[2], char *fullSocket)
+{
+    //In the child pam_syslog does not work, using syslog directly
+
+    //keep stderr open so socket doesn't returns us that fd
+    int x = 3;
+    //Set FD_CLOEXEC on fd that are not of interest of kwallet
+    for (; x < 64; ++x) {
+        if (x != toWalletPipe[0]) {
+            fcntl(x, F_SETFD, FD_CLOEXEC);
+        }
+    }
```

Comment:
Trying to get rid of the lowest 64 file descriptors this way is both inefficient and incomplete. The clean way to do this would be looking into /proc/self/fd and operate only on the files that are actually open. At least this loop should look at all 1024 possible file descriptors then.

Reply:
An even better way would be to use the `close_range` system call.